config_types.cpp

```diff
@@ -263,3 +266,93 @@ const char* server_addr::get_last_error(void) const
 {
     return gai_strerror(m_last_error);
 }
+
+config_cpu_list::config_cpu_list(unsigned int cores)
+{
+    for (unsigned int i = 0; i < cores; ++i)
+        cpu_list.push_back(i);
+
+    next_cpu = cpu_list.begin();
+}
+
+config_cpu_list::config_cpu_list(std::string str)
+{
+    std::stringstream ss(str);
+
+    if (str.find(' ') != std::string::npos)
+        return;
+
+    char separator = str[str.find_first_of(",-")];
+
+    int i;
+    while (ss >> i)
+    {
+        cpu_list.push_back(i);
+
+        if (ss.peek() == separator)
+            ss.ignore();
+        else if (ss.eof())
+            break;
+        else {
+            cpu_list.clear();
+            break;
+        }
+    }
+
+    next_cpu = cpu_list.begin();
+}
+
+config_cpu_list::config_cpu_list(const config_cpu_list& copy) :
+        cpu_list(copy.cpu_list)
+{
+    next_cpu = cpu_list.begin();
+}
+
+config_cpu_list& config_cpu_list::operator=(const config_cpu_list& cl)
+{
+    if (this == &cl)
+        return *this;
+
+    cpu_list = cl.cpu_list;
+    next_cpu = cpu_list.begin();
+    return *this;
+}
+
+bool config_cpu_list::is_defined() const
```

Comment:
why do you need this function? why not failing the compiler if you can't create a valid cpu list?

Reply:
it is used to determine whether to initialize the taskset to with a default list of cpus. it is also used to determine whether to use affinity or not